RiotSwiftUI/Modules/Authentication/Common/Service/MatrixSDK/AuthenticationService.swift

```diff
@@ -0,0 +1,334 @@
+    }
+    
+    
+    // MARK: Android OnboardingViewModel
+    /// The current state of the authentication flow.
+    private var state = AuthenticationCoordinatorState()
+    /// The currently executing async task.
+    private var currentTask: Task<Void, Error>? {
+        willSet {
+            currentTask?.cancel()
+        }
+    }
+    
+    
+    // MARK: - Setup
+    
+    override init() {
+        guard let homeserverURL = URL(string: RiotSettings.shared.homeserverUrlString) else {
+            fatalError("[AuthenticationService]: Failed to create URL from default homeserver URL string.")
+        }
+        
+        client = MXRestClient(homeServer: homeserverURL, unrecognizedCertificateHandler: nil)
+        
+        super.init()
+    }
+    
+    // MARK: - Android OnboardingViewModel
+    
+    func loginFlow(homeserverAddress: String) async {
+        currentTask = Task {
+            cancelPendingLoginOrRegistration()
+            
+            do {
+                let data = try await loginFlow(for: homeserverAddress)
+                
+                guard !Task.isCancelled else { return }
+                
+                // Valid Homeserver, add it to the history.
+                // Note: we add what the user has input, as the data can contain a different value.
+                RiotSettings.shared.homeserverUrlString = homeserverAddress
+                
+                let loginMode: LoginMode
+                
+                if data.supportedLoginTypes.contains(where: { $0.type == kMXLoginFlowTypeSSO }),
+                   data.supportedLoginTypes.contains(where: { $0.type == kMXLoginFlowTypePassword }) {
+                    loginMode = .ssoAndPassword(ssoIdentityProviders: data.ssoIdentityProviders)
+                } else if data.supportedLoginTypes.contains(where: { $0.type == kMXLoginFlowTypeSSO }) {
+                    loginMode = .sso(ssoIdentityProviders: data.ssoIdentityProviders)
+                } else if data.supportedLoginTypes.contains(where: { $0.type == kMXLoginFlowTypePassword }) {
+                    loginMode = .password
+                } else {
+                    loginMode = .unsupported
+                }
+                
+                state.homeserverAddressFromUser = homeserverAddress
+                state.homeserverAddress = data.homeserverAddress
+                state.loginMode = loginMode
+                state.loginModeSupportedTypes = data.supportedLoginTypes
+            } catch {
+                #warning("Show an error message and/or handle the error?")
+                return
+            }
+        }
+    }
+    
+    func refreshServer(homeserverAddress: String) async throws -> (LoginFlowResult, RegistrationResult) {
+        let loginFlows = try await loginFlow(for: homeserverAddress)
+        let wizard = try registrationWizard()
+        let registrationFlow = try await wizard.registrationFlow()
```

Comment:
Hmmm this is a nice idea, although it wouldn't work right now due to the `registrationWizard()` expecting the `loginFlow` call to have finished so the REST client has been configured based on the well-known data (and the `pendingData` is ready for use).

Reply:
In that case I have two further questions:

- why does anything called "registration" depend on anything called "login", shouldn't they be either independent flows, or perhaps called differently? (e.g. login is really something like authenticationSetup)?
- if they do indeed depend on each other, it would be better to use the output of the first function as input of the other. At the moment someone could randomly swap around the lines without knowing the order matters.

Reply:
> if they do indeed depend on each other, it would be better to use the output of the first function as input of the other. At the moment someone could randomly swap around the lines without knowing the order matters.

The `pendingData` part is now fixed. I like the idea of pulling the well-known part up one level though and passing it in. Again will add it to my list 👍

> why does anything called "registration" depend on anything called "login", shouldn't they be either independent flows, or perhaps called differently? (e.g. login is really something like authenticationSetup)?

Ignoring the concepts I've brought over from Android, as far as I can see the single reason from a higher level is that the Registration page also shows any SSO buttons and those are actually discovered through the login flow. So we will definitely need to set up a login flow alongside a registration flow in order to get those. But I think I can see how this can be better structured now :)